core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -73,57 +76,114 @@ private void analyze(ActivityPayload payload, UserAnonymizedDTO userAnonymized,
-		ActivityPayload nextDayPayload = null;
 		if (isCrossDayActivity(payload, userAnonymized))
 		{
-			// crossing day border
-			// set current payload to end of day
-			// use new payload for remainder
 			// assumption: activity never crosses 2 days
-			truncatedPayload = new ActivityPayload(payload.url, payload.startTime,
+			ActivityPayload truncatedPayload = new ActivityPayload(payload.url, payload.startTime,
 					Date.from(getEndOfDay(payload.startTime, userAnonymized).toInstant()), payload.application);
 
-			nextDayPayload = new ActivityPayload(payload.url,
+			ActivityPayload nextDayPayload = new ActivityPayload(payload.url,
 					Date.from(getStartOfDay(payload.endTime, userAnonymized).toInstant()), payload.endTime, payload.application);
-		}
-
-		addOrUpdateDayTruncatedActivity(truncatedPayload, userAnonymized, matchingGoal);
 
-		if (nextDayPayload != null)
+			addOrUpdateDayTruncatedActivity(truncatedPayload, userAnonymized, matchingGoal);
+			addOrUpdateDayTruncatedActivity(nextDayPayload, userAnonymized, matchingGoal);
+		}
+		else
 		{
-			// day should be created
-			addActivity(nextDayPayload, userAnonymized, matchingGoal, null);
+			addOrUpdateDayTruncatedActivity(payload, userAnonymized, matchingGoal);
 		}
 	}
 
 	private void addOrUpdateDayTruncatedActivity(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Goal matchingGoal)
 	{
-		Date minEndTime = new Date(payload.startTime.getTime() - yonaProperties.getAnalysisService().getConflictInterval());
-		DayActivity dayActivity = getCachedDayActivity(payload, userAnonymized, matchingGoal);
-		Activity activity = dayActivity == null ? null : dayActivity.getLastActivity();
-		if (activity == null || activity.getEndTime().before(minEndTime))
+		DayActivityCacheResult dayActivity = getRegisteredDayActivity(payload, userAnonymized, matchingGoal);
+		if (canCombineWithLastRegisteredActivity(payload, lastRegisteredActivity))
+		{
+			if (isBeyondSkipWindowAfterLastRegisteredActivity(payload, lastRegisteredActivity))
+			{
+				// Update message only if it is within five seconds to avoid unnecessary cache flushes.
+				updateActivityEndTime(payload, userAnonymized, matchingGoal, dayActivity, lastRegisteredActivity);
+			}
+		}
+		else
 		{
 			addActivity(payload, userAnonymized, matchingGoal, dayActivity);
 		}
-		// Update message only if it is within five seconds to avoid unnecessary cache flushes.
-		// Or if the start time is earlier than the existing start time (this can happen with app activity, see below).
-		else if (payload.endTime.getTime() - activity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
-				.getUpdateSkipWindow() || payload.startTime.before(activity.getStartTime()))
+	}
+
+	private boolean isBeyondSkipWindowAfterLastRegisteredActivity(ActivityPayload payload, Activity lastRegisteredActivity)
+	{
+		return payload.endTime.getTime() - lastRegisteredActivity.getEndTime().getTime() >= yonaProperties.getAnalysisService()
```

Comment:
Shouldn't you compare the start time of the new activity with the end time of the last registered one?